x/warp/contract.go

```diff
@@ -0,0 +1,278 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	predicateutils "github.com/ava-labs/subnet-evm/utils/predicate"
+	"github.com/ava-labs/subnet-evm/vmerrs"
+	warpPayload "github.com/ava-labs/subnet-evm/warp/payload"
+
+	_ "embed"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+)
+
+const (
+	GetBlockchainIDGasCost uint64 = 2 // Based on GasQuickStep used in existing EVM instructions
+	// Sum of base log gas cost, cost of producing 4 topics, and producing + serving a BLS Signature (sign + trie write)
+	// Note: using trie write for the gas cost results in a conservative overestimate since the message is stored in a
+	// flat database that can be cleaned up after a period of time instead of the EVM trie.
+	SendWarpMessageGasCost uint64 = params.LogGas + 4*params.LogTopicGas + 20_000 + contract.WriteGasCostPerSlot
+	// SendWarpMessageGasCostPerByte cost accounts for producing a signed message of a given size
+	SendWarpMessageGasCostPerByte uint64 = params.LogDataGas
+
+	GasCostPerWarpSigner            uint64 = 500
+	GasCostPerWarpMessageBytes      uint64 = 100 // TODO: charge O(n) cost for decoding predicate of input size n
+	GasCostPerSignatureVerification uint64 = 200_000
+	// GasCostPerSourceSubnetValidator uint64 = 1 // TODO: charge O(n) cost for subnet validator set lookup
+)
+
+var (
+	errInvalidSendInput = errors.New("invalid sendWarpMessage input")
+)
+
+// Singleton StatefulPrecompiledContract and signatures.
+var (
+
+	// WarpRawABI contains the raw ABI of Warp contract.
+	//go:embed contract.abi
+	WarpRawABI string
+
+	WarpABI = contract.ParseABI(WarpRawABI)
+
+	WarpPrecompile = createWarpPrecompile()
+)
+
+// WarpMessage is an auto generated low-level Go binding around an user-defined struct.
+type WarpMessage struct {
+	OriginChainID       [32]byte
+	OriginSenderAddress [32]byte
+	DestinationChainID  [32]byte
+	DestinationAddress  [32]byte
```

Comment:
It's just an alias, but this is what was produced by the precompile template. If we want to use `common.Hash` instead of `[32]byte` we should update the precompile template to handle the special case. Would you prefer I update these to use `common.Hash` in the meantime?

Reply:
If it also makes sense to you, we can both convert here and in template. I can change the template in another PR.

Reply:
Updated to `common.Hash`